File: R/volume.R
#' Calculate volume
#'
#' Calculates volume from taper model by integrating the taper function.
#'
#' @param dbh numeric vector of diameters at breast height, 1.3 m above ground (cm).
#' @param h_top numeric vector of tree heights above ground (m).
#' @param h_vol_lower,h_vol_upper numeric vectors of heights between stem volume is calculated (m). Default stump height (h_vol_lower) is 1% of h_top. NA in h_vol_upper is replaced by h_top.
#' @param sp species.
#' @param with_bark calculate volume over (TRUE, default) or under bark (FALSE).
#' @return Timber volume (m^3).
#' @examples
#' volume(20, 30)
#' volume(dbh=c(20,25,30), h_top=c(30,25,37))
#' volume(dbh=rep(25,11), h_top=rep(30,11),h_vol_lower=seq(0,30,3))
#' volume(dbh=rep(25,11), h_top=rep(30,11),h_vol_upper=seq(0,30,3))
#' volume(dbh=rep(25,11), h_top=rep(30,11),h_vol_upper=seq(0,30,3),with_bark=FALSE)
#' @export


volume<-function(dbh,h_top,h_vol_lower=NA,h_vol_upper=NA,sp="spruce",with_bark=TRUE){

  if(class(dbh)!="numeric"|class(h_top)!="numeric"|class(h_vol_lower)!="numeric"){
    stop("dbh, h_top and h_vol_lower must be numeric.")
  }

  if(length(dbh)!=length(h_top)) {
    stop("dbh and h_top must have the same length.")
  }

  if(length(h_vol_upper)==1) {
    h_vol_upper<-rep(h_vol_upper,length(dbh))
  } else if (length(h_vol_upper)!=length(dbh)){
    stop("h_vol_upper must be of length 1 of same length as dbh.")
  }

  #replace NA with h_top
  h_vol_upper[is.na(h_vol_upper)]<-h_top[is.na(h_vol_upper)]

  if(length(h_vol_lower)==1) {
      h_vol_lower<-rep(h_vol_lower,length(dbh))
  } else if (length(h_vol_lower)!=length(dbh)){
    stop("h_vol_lower must be of length 1 of same length as dbh.")
  }
  
  #replace NA with 0.1*h_top
  h_vol_lower[is.na(h_vol_lower)]<-0.1*h_top[is.na(h_vol_lower)]
  
  if(any(h_vol_lower>h_vol_upper)){
    stop("h_vol_lower must not be larger than h_vol_upper.")
  }

  if(length(sp)==1) {
    sp<-rep(sp,length(dbh))
  } else if (length(sp)!=length(dbh)){
    stop("sp must be of length 1 of same length as dbh.")
  }

  if(length(with_bark)==1) {
    with_bark<-rep(with_bark,length(dbh))
  } else if (length(with_bark)!=length(dbh)){
    stop("with_bark must be of length 1 of same length as dbh.")
  }


  taper_integr<-unlist(apply( data.frame(dbh,h_top,h_vol_lower,h_vol_upper,sp,with_bark),
                              MARGIN = 1,
                              FUN= function(x){
                                stats::integrate(function(h,dbh,h_top,sp,with_bark)((taperNOR(h,dbh,h_top,sp,with_bark)/100)/2)^2,
                                                 dbh=as.numeric(x[1]),
                                                 h_top=as.numeric(x[2]),
                                                 sp=x[5],
                                                 with_bark=as.logical(x[6]),
                                                 lower = as.numeric(x[3]),
                                                 upper = as.numeric(x[4]))$value
                              }
  ))



  return(pi*unname(taper_integr))
}

File: python/taper_nor/volume.py
"""Stem volume by integration of the taper curve."""

from __future__ import annotations

import math
from typing import Sequence

import numpy as np
from scipy import integrate

from .coefficients import normalize_species
from .taper import _taper


def volume(
    dbh: Sequence[float],
    h_top: Sequence[float],
    h_vol_lower: Sequence[float | None] | None = None,
    h_vol_upper: Sequence[float | None] | None = None,
    sp: str = "spruce",
    with_bark: bool = True,
) -> list[float]:
    """Stem volume (m^3) by integrating the taper curve.

    Args:
        dbh: Diameter at breast height (cm), one entry per tree.
        h_top: Total tree height (m), one entry per tree.
        h_vol_lower: Lower integration height (m) per tree. Defaults to 1% of
            ``h_top``. ``None`` entries fall back to this default.
        h_vol_upper: Upper integration height (m) per tree. Defaults to
            ``h_top``. ``None`` entries fall back to this default.
        sp: Species, see :func:`taper_nor.coefficients.normalize_species`.
        with_bark: Integrate diameter over bark (``True``, default) or under
            bark (``False``).

    Returns:
        Stem volume (m^3) for each tree.

    Raises:
        ValueError: If ``dbh``, ``h_top``, ``h_vol_lower``, or ``h_vol_upper``
            do not all have the same length.
    """
    n = len(dbh)
    if len(h_top) != n:
        raise ValueError("dbh and h_top must have the same length")

    sp_n = normalize_species(sp)

    lowers = h_vol_lower if h_vol_lower is not None else [None] * n
    uppers = h_vol_upper if h_vol_upper is not None else [None] * n
    if len(lowers) != n or len(uppers) != n:
        raise ValueError("h_vol_lower and h_vol_upper must be the same length as dbh")

    results = []
    for d, ht, lo, up in zip(dbh, h_top, lowers, uppers):
        upper = ht if up is None else up
        lower = ht * 0.01 if lo is None else lo
        if lower > upper:
            lower = upper

        def integrand(h: float, d: float = d, ht: float = ht) -> float:
            taper = _taper(np.array([h]), d, ht, sp_n, with_bark)[0]
            return ((taper / 100) / 2) ** 2

        val, _ = integrate.quad(integrand, lower, upper)
        results.append(round(math.pi * val, 6))

    return results
